### DISCUSSION TITLE
採番機能について任意の月を開始月とする年度でローテーションできる機能が欲しい

### DISCUSSION THREAD

**Discussion by koichimurakami** [mbc-net/mbc-cqrs-serverless#37] [Ideas]
現状 sequence パッケージにおいて rotate_by に指定出来る値として fiscal_yearly があるが4月開始となっている。
現状日本の法人では事業年度の開始日は任意であるため、採番機能としても任意の日付から事業年度を計算してほしい。

**Comment by koichimurakami** [comment]
事業年度を〇〇期で返却したいケースもあるので、2024年度=2024年4月〜2025年3月だけではなく、第20期=2024年4月〜2025年3月といった計算をしたい。第1期は会社設立初年度等を想定しているので、企業毎に日付から任意の数値を返却できるように関数を設定出来るのが良いと思う。

**Comment by ghost** [comment]
@koichimurakami 
以下のインターフェースを使用して、シーケンスサービスを実装する予定です。
```
export interface ISequenceService {
  /**
   * Get the current sequence by a specific key.
   * @param key - The key to identify the sequence details.
   * @returns A promise that resolves to the current sequence's data entity.
   */
  getCurrentSequence(key: DetailKey): Promise<DataEntity>

  /**
   * Generate a new sequence based on the provided parameters.
   * @param dto - The data transfer object containing generation parameters.
   * @param opts - Additional options including invocation context.
   * @returns A promise that resolves to the newly generated sequence's data entity.
   */
  genNewSequence(
    dto: GenSequenceDto,
    opts: {
      invokeContext: IInvoke
    },
  ): Promise<DataEntity>

  /**
   * Generate a new sequence with a specified format.
   * @param dto - The data transfer object containing parameters for formatted sequence generation.
   * @param opts - Additional options including invocation context.
   * @returns A promise that resolves to the newly generated formatted sequence's data entity.
   */
  generateFormattedSequence(
    dto: GenerateFormattedSequenceDto,
    opts: {
      invokeContext: IInvoke
    },
  ): Promise<DataEntity>
}
```
最初の2つのメソッドは以前のバージョンで既に実装されており、generateFormattedSequenceは新しい関数になります。

**Reply by koichimurakami** [reply]
@quangquandev 
GenSequenceDto と GenerateFormattedSequenceDto の定義も教えて頂けますか？

上2つは維持する予定ですか？
上2つを統合して generateSequenceItem として、採番された番号、フォーマット済みの番号をオブジェクト型として返却する関数を作ることをイメージしていましたが、採番した番号だけと採番してフォーマットした番号を返却したい場合どちらも対応できるのであればそれでも構いません。
私が開発していたときにはフォーマット済みの番号と採番した番号だけのものの両方が必要になったことがあります。

**Reply by ghost** [reply]
@koichimurakami 
Service で以前から定義されている getCurrentSequence と genNewSequence の 2 つの項目は全く使用されていないため、Interface から削除したいと思います。

**Reply by ghost** [reply]
@koichimurakami 
以下のGenSequenceDto と GenerateFormattedSequenceDto の定義です。
```
export class GenSequenceDto {
  /**
   * Date for sequence generation (optional).
   */
  @IsOptional()
  @Type(() => Date)
  @ApiProperty({
    type: Date,
    required: false,
    description: 'Date for sequence generation.',
  })
  date?: Date

  /**
   * Rotation criteria for the sequence (e.g., yearly, fiscal-yearly).
   */
  @ApiProperty({ enum: RotateByEnum, example: RotateByEnum.FISCAL_YEARLY })
  @IsOptional()
  @IsEnum(RotateByEnum)
  rotateBy?: RotateByEnum

  /**
   * Tenant code for identifying the organization.
   */
  @IsString()
  @ApiProperty({ description: 'Code of the tenant (organization).' })
  tenantCode: string

  /**
   * Type code for specific sequence classification.
   */
  @IsString()
  @ApiProperty({
    required: true,
    description: 'Type code for specific classification.',
  })
  typeCode: string
}

export class GenerateFormattedSequenceDto {
  /**
   * Date for sequence generation (optional).
   */
  @IsOptional()
  @Type(() => Date)
  @ApiProperty({
    type: Date,
    required: false,
    description: 'Date for sequence generation.',
  })
  date?: Date

  /**
   * Rotation criteria for the sequence (e.g., yearly, fiscal-yearly).
   */
  @ApiProperty({ enum: RotateByEnum, example: RotateByEnum.FISCAL_YEARLY })
  @IsOptional()
  @IsEnum(RotateByEnum)
  rotateBy?: RotateByEnum

  /**
   * Tenant code for identifying the organization.
   */
  @IsString()
  tenantCode: string

  /**
   * Type code for specific sequence classification.
   */
  @IsString()
  typeCode: string

  @IsObject()
  params: SequenceParamsDto

  /**
   * Format of the sequence number.
   */
  @IsString()
  @IsOptional()
  @ApiProperty({
    required: false,
    description: 'Format of the sequence number (e.g., ABC-###).',
  })
  format?: string

  /**
   * Start month of fiscal year (default: 4)
   */

  @IsNumber()
  @IsOptional()
  @ApiProperty({
    required: false,
    description: 'Start month of fiscal year. Default is 4.',
    default: 4,
  })
  startMonth?: number

  /**
   * Company registration date
   */

  @IsOptional()
  @ApiProperty({
    type: Date,
    required: false,
    description: 'Company registration date.',
  })
  registerDate?: Date
}
```

**Reply by koichimurakami** [reply]
@quangquandev 
上2つを削除するのであれば V0.2まではDeprecatedにしておいて下さい。
残るのは `generateFormattedSequence` だけであれば `generateSequnceItem` として、採番したデータをオブジェクトとして返却してほしいです。オブジェクトの中身としては `no` 、`formatted_no` を返却してください。

## `GenerateFormattedSequenceDto` について
* `typeCode` に基づき採番をする想定ですが、マスタ設定があれば番号のフォーマット、開始年月のデータを取得する想定ですが合っていますか？
* `params` は任意に出来ませんか？ `SequenceParamsDto ` の定義もみたいです。
* ` format` については `typeCode` から取得するマスタ設定にフォーマットがあればそれを取得する想定なので不要だと思っています。
* `startMonth` も`typeCode` から取得するマスタ設定(テナント毎に異なる可能性もある)に開始月がある想定ですがいかがでしょうか？

ちなみに以下の対応時に `tenantCode` は任意にして貰いたいと考えています。
https://github.com/mbc-net/mbc-cqrs-serverless/discussions/50

**Reply by ghost** [reply]
@koichimurakami 

- Deprecatedに関して、今処理します。
- 返却しるデータにno や formatted_noなど は、返されるデータのattributesに含まれました。
**GenerateFormattedSequenceDto について**
以前、シーケンスの部分を柔軟にするために、マスターデータの取得は関数の外で実行していましたが、今はその処理を関数内に移し、引数の渡し方も変更します。

**Reply by koichimurakami** [reply]
@quangquandev 
> 返却しるデータにno や formatted_noなど は、返されるデータのattributesに含まれました。

返却データの定義も見せてもらえますか？

> 以前、シーケンスの部分を柔軟にするために、マスターデータの取得は関数の外で実行していましたが、今はその処理を関数内に移し、引数の渡し方も変更します。

本件ですが、 マスタ関連をパッケージ化したいと考えています。
マスタパッケージを注入したらそこからデータを取得し、そうでなければ関数から取得する事って出来ますか？
関数を上書きすれば動作も変更出来れば良いのかなと考えていますが、いかがでしょうか？

**Reply by ghost** [reply]
@koichimurakami 
こちらの返却データの定義です。
```
 const mockResponse = {
        code: "TODO#TASK1#S#2024",
        updatedBy: "92ca4f68-9ac6-4080-9ae2-2f02a86206a4",
        createdIp: "127.0.0.1",
        tenantCode: tenantCode,
        type: "TODO",
        createdAt: "2024-11-08T13:50:26+07:00",
        updatedIp: "127.0.0.1",
        createdBy: "92ca4f68-9ac6-4080-9ae2-2f02a86206a4",
        requestId: "81bf1821-34b0-4dc5-a2ce-685d37d22f8c",
        name: "fiscal_yearly",
        sk: "TODO#TASK1#S＃2024",
        attributes: {
          fiscal_year: 71,
          issued_at: "2024-11-08T13:50:26+07:00",
          formatted_no: "00TASK1-71-S-001",
          no:1
        },
        pk: "SEQ#MBC",
        seq: 1,
        updatedAt: "2024-11-08T13:50:26+07:00"
      }
```

**Reply by ghost** [reply]
> @quangquandev

> 本件ですが、 マスタ関連をパッケージ化したいと考えています。 マスタパッケージを注入したらそこからデータを取得し、そうでなければ関数から取得する事って出来ますか？ 関数を上書きすれば動作も変更出来れば良いのかなと考えていますが、いかがでしょうか？

それなら、シーケンスで使用できるように、マスターパッケージをmbc-framwork内に作成する必要があります。
https://github.com/mbc-net/mbc-cqrs-serverless/discussions/52

**Reply by koichimurakami** [reply]
@quangquandev 
返却データですが、DynamoDBのデータをそのまま返却するのは止めましょう。 `seq` が採番された数値だと勘違いする可能性があります。

下記の様な返却データはいかがですか？

```
{
  "id": "採番のID DynamoDBのID(文字列型)",
  "no": "採番された番号(数値型)",
  "formattedNo": "フォーマットされた番号(文字列型)",
  "issuedAt": "採番した日時または、基準日時"
}
```

**Reply by koichimurakami** [reply]
@quangquandev 
> それなら、シーケンスで使用できるように、マスターパッケージをmbc-framwork内に作成する必要があります。

いえ、これはどうしてもパッケージにしたいです。パッケージをインストールしている場合としていない場合の両方をシーケンス側の実装をする必要があるかもしれないです。
どちらの場合も同じメソッドで同じインターフェースを使うように設計してほしいです。
インターフェースのみCoreにあっても良いかもしれません。

**Reply by ghost** [reply]
@koichimurakami 
お疲れ様です
typeCodeとcode1〜5の合計6個指定出来るように修正しました。
このプルを確認お願い致します
https://github.com/mbc-net/mbc-cqrs-serverless/pull/78

**Comment by koichimurakami** [comment]
Document: https://mbc-cqrs-serverless.mbc-net.com/docs/sequence/
Document in Japanese: https://mbc-cqrs-serverless.mbc-net.com/ja/docs/sequence/
